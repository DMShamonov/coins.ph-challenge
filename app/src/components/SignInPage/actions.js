import api from 'controllers/api';
import { setAccessToken } from 'controllers/session';
import { fetchProfile } from 'controllers/profile';

import ROUTES from 'constants/routes';
import SIGN_IN_FORM_ACTION_TYPES from 'redux/actionTypes/signInForm';
import SESSION_ACTION_TYPES from 'redux/actionTypes/session';

export function signIn(history) { // eslint-disable-line import/prefer-default-export
  return async function _signIn(dispatch, getState) {
    dispatch({ type: SIGN_IN_FORM_ACTION_TYPES.ON_START_PROCESSING });

    const { ui: { signInForm: { login, password } } } = getState();

    try {
      const { access_token } = await api().authenticate({ username: login, password, otp: '0000000' });

      setAccessToken(access_token);

      const profile = await fetchProfile();

      dispatch({ type: SESSION_ACTION_TYPES.LOGIN, payload: { profile } });

      history.push(ROUTES.ROOT);

      dispatch({ type: SIGN_IN_FORM_ACTION_TYPES.ON_RESET_FORM });
    } catch (e) {
      const error = e?.result?.message || 'An error has been occurred. Try again, please.';

      dispatch({ type: SIGN_IN_FORM_ACTION_TYPES.ON_SET_ERROR, error });
    }

    dispatch({ type: SIGN_IN_FORM_ACTION_TYPES.ON_FINISH_PROCESSING });
  };
}
