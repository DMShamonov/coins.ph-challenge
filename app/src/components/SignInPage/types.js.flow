// @flow

export type PropsType = {|
  isLogged: boolean,
  login: string,
  password: string,
  error: string,
  hasError: boolean,
  isFormDisabled: boolean,
  onChangeLogin: SyntheticInputEvent<HTMLInputElement> => void,
  onChangePassword: SyntheticInputEvent<HTMLInputElement> => void,
  onSubmit: Event => void,
|};
