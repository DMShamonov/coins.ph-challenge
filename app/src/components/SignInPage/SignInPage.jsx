// @flow

import * as React from 'react';
import { Redirect } from 'react-router-dom';

import ROUTES from 'constants/routes';
import { Form, FormGroup, FormError } from 'components/Form';
import Input from 'components/Input';
import Button from 'components/Button';

import type { PropsType } from './types.js.flow';

export default function SignInPage({
  isLogged,
  login,
  password,
  error,
  hasError,
  isFormDisabled,
  onChangeLogin,
  onChangePassword,
  onSubmit,
}: PropsType): React.Node {
  if (isLogged) {
    return <Redirect to={{ pathname: ROUTES.ROOT }} />;
  }

  return (
    <div className="sign-in">
      <div className="sign-in__content">
        <h1 className="sign-in__title">Sign in to BitGo account</h1>
        <Form onSubmit={onSubmit}>
          <FormGroup>
            <Input
              type="email"
              name="login"
              label="Login"
              placeholder="Enter your BitGo login"
              value={login}
              autoFocus
              required
              disabled={isFormDisabled}
              onChange={onChangeLogin}
            />
          </FormGroup>
          <FormGroup>
            <Input
              type="password"
              name="password"
              label="Password"
              placeholder="Enter your BitGo password"
              value={password}
              autoComplete="off"
              required
              disabled={isFormDisabled}
              onChange={onChangePassword}
            />
          </FormGroup>
          {hasError && <FormError>{error}</FormError>}
          <Button type="submit" disabled={isFormDisabled}>Sign in</Button>
        </Form>
      </div>
    </div>
  );
}
