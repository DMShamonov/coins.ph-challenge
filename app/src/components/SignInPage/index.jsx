import { compose } from 'redux';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';

import verifySession from 'decorators/verifySession';
import SIGN_IN_FORM_ACTION_TYPES from 'redux/actionTypes/signInForm';

import { signIn } from './actions';
import SignInPage from './SignInPage';

function mapStateToProps({ ui: { signInForm } }) {
  const { isProcessing, error, ...formData } = signInForm;
  const hasError = !!error;

  return {
    ...formData,
    error,
    hasError,
    isFormDisabled: isProcessing,
  };
}

function mapDispatchToProps(dispatch, { history }) {
  return {
    onChangeLogin({ target: { value: login } }) {
      dispatch({
        type: SIGN_IN_FORM_ACTION_TYPES.ON_CHANGE_LOGIN,
        payload: { login },
      });
    },
    onChangePassword({ target: { value: password } }) {
      dispatch({
        type: SIGN_IN_FORM_ACTION_TYPES.ON_CHANGE_PASSWORD,
        payload: { password },
      });
    },
    onSubmit(e) {
      e.preventDefault();

      dispatch(signIn(history));
    },
  };
}

export default compose(
  withRouter,
  connect(
    mapStateToProps,
    mapDispatchToProps,
  ),
  verifySession,
)(SignInPage);
