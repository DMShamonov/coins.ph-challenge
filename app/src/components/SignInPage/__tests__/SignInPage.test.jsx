import React from 'react';
import { shallow, mount } from 'enzyme';

import SignInPage from '../SignInPage';

describe('`SignInPage` component should render correctly', () => {
  test('with default props', () => {
    expect(shallow((
      <SignInPage
        isLogged={false}
        login=""
        password=""
        error=""
        hasError={false}
        isFormDisabled={false}
        onChangeLogin={jest.fn()}
        onChangePassword={jest.fn()}
        onSubmit={jest.fn()}
      />
    ))).toMatchSnapshot();
  });

  test('when `isLogged` prop is `true`', () => {
    expect(shallow((
      <SignInPage
        isLogged
        login=""
        password=""
        error=""
        hasError={false}
        isFormDisabled={false}
        onChangeLogin={jest.fn()}
        onChangePassword={jest.fn()}
        onSubmit={jest.fn()}
      />
    ))).toMatchSnapshot();
  });

  test('when `hasError` prop is `true` and `error` prop has value', () => {
    expect(shallow((
      <SignInPage
        isLogged={false}
        login=""
        password=""
        error="__ERROR__"
        hasError
        isFormDisabled={false}
        onChangeLogin={jest.fn()}
        onChangePassword={jest.fn()}
        onSubmit={jest.fn()}
      />
    ))).toMatchSnapshot();
  });

  test('when `isFormDisabled` prop is `true`', () => {
    expect(shallow((
      <SignInPage
        isLogged={false}
        login=""
        password=""
        error=""
        hasError={false}
        isFormDisabled
        onChangeLogin={jest.fn()}
        onChangePassword={jest.fn()}
        onSubmit={jest.fn()}
      />
    ))).toMatchSnapshot();
  });
});

describe('Check `SignInPage` props', () => {
  describe('Check callback props', () => {
    test('`onChangeLogin` called', () => {
      const onChangeLogin = jest.fn();
      const wrapper = mount((
        <SignInPage
          isLogged={false}
          login=""
          password=""
          error=""
          hasError={false}
          isFormDisabled={false}
          onChangeLogin={onChangeLogin}
          onChangePassword={jest.fn()}
          onSubmit={jest.fn()}
        />
      ));

      wrapper.find('input[name="login"]').simulate('change');

      expect(onChangeLogin).toHaveBeenCalled();
    });

    test('`onChangePassword` called', () => {
      const onChangePassword = jest.fn();
      const wrapper = mount((
        <SignInPage
          isLogged={false}
          login=""
          password=""
          error=""
          hasError={false}
          isFormDisabled={false}
          onChangeLogin={jest.fn()}
          onChangePassword={onChangePassword}
          onSubmit={jest.fn()}
        />
      ));

      wrapper.find('input[name="password"]').simulate('change');

      expect(onChangePassword).toHaveBeenCalled();
    });

    test('`onSubmit` called', () => {
      const onSubmit = jest.fn();
      const wrapper = mount((
        <SignInPage
          isLogged={false}
          login=""
          password=""
          error=""
          hasError={false}
          isFormDisabled={false}
          onChangeLogin={jest.fn()}
          onChangePassword={jest.fn()}
          onSubmit={onSubmit}
        />
      ));

      wrapper.find('form').simulate('submit');

      expect(onSubmit).toHaveBeenCalled();
    });
  });
});
