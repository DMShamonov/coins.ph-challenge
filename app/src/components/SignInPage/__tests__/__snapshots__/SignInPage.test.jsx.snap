// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`SignInPage\` component should render correctly when \`hasError\` prop is \`true\` and \`error\` prop has value 1`] = `
<div
  className="sign-in"
>
  <div
    className="sign-in__content"
  >
    <h1
      className="sign-in__title"
    >
      Sign in to BitGo account
    </h1>
    <Form
      onSubmit={[MockFunction]}
    >
      <FormGroup>
        <Input
          autoFocus={true}
          label="Login"
          name="login"
          onChange={[MockFunction]}
          placeholder="Enter your BitGo login"
          required={true}
          type="email"
          value=""
        />
      </FormGroup>
      <FormGroup>
        <Input
          autoComplete="off"
          label="Password"
          name="password"
          onChange={[MockFunction]}
          placeholder="Enter your BitGo password"
          required={true}
          type="password"
          value=""
        />
      </FormGroup>
      <FormError>
        __ERROR__
      </FormError>
      <Button
        disabled={false}
        type="submit"
      >
        Sign in
      </Button>
    </Form>
  </div>
</div>
`;

exports[`\`SignInPage\` component should render correctly when \`isFormDisabled\` prop is \`true\` 1`] = `
<div
  className="sign-in"
>
  <div
    className="sign-in__content"
  >
    <h1
      className="sign-in__title"
    >
      Sign in to BitGo account
    </h1>
    <Form
      onSubmit={[MockFunction]}
    >
      <FormGroup>
        <Input
          autoFocus={true}
          label="Login"
          name="login"
          onChange={[MockFunction]}
          placeholder="Enter your BitGo login"
          required={true}
          type="email"
          value=""
        />
      </FormGroup>
      <FormGroup>
        <Input
          autoComplete="off"
          label="Password"
          name="password"
          onChange={[MockFunction]}
          placeholder="Enter your BitGo password"
          required={true}
          type="password"
          value=""
        />
      </FormGroup>
      <Button
        disabled={true}
        type="submit"
      >
        Sign in
      </Button>
    </Form>
  </div>
</div>
`;

exports[`\`SignInPage\` component should render correctly when \`isLogged\` prop is \`true\` 1`] = `
<Redirect
  to={
    Object {
      "pathname": "/",
    }
  }
/>
`;

exports[`\`SignInPage\` component should render correctly with default props 1`] = `
<div
  className="sign-in"
>
  <div
    className="sign-in__content"
  >
    <h1
      className="sign-in__title"
    >
      Sign in to BitGo account
    </h1>
    <Form
      onSubmit={[MockFunction]}
    >
      <FormGroup>
        <Input
          autoFocus={true}
          label="Login"
          name="login"
          onChange={[MockFunction]}
          placeholder="Enter your BitGo login"
          required={true}
          type="email"
          value=""
        />
      </FormGroup>
      <FormGroup>
        <Input
          autoComplete="off"
          label="Password"
          name="password"
          onChange={[MockFunction]}
          placeholder="Enter your BitGo password"
          required={true}
          type="password"
          value=""
        />
      </FormGroup>
      <Button
        disabled={false}
        type="submit"
      >
        Sign in
      </Button>
    </Form>
  </div>
</div>
`;
